mesonbuild/scripts/depfixer.py

```diff
@@ -374,7 +376,26 @@ def fix_darwin(fname, new_rpath, final_path, install_name_mappings):
     try:
         args = []
         if rpaths:
-            for rp in rpaths:
+            # TODO: fix this properly, not totally clear how
+            #
+            # removing rpaths from binaries on macOS has tons of
+            # weird edge cases. For instance, if the user provided
+            # a '-Wl,-rpath' argument in LDFLAGS that happens to
+            # coincide with an rpath generated from a dependency,
+            # this would cause installation failures, as meson would
+            # generate install_name_tool calls with two identical
+            # '-delete_rpath' arguments, which install_name_tool
+            # fails on. Because meson itself ensures that it never
+            # adds duplicate rpaths, duplicate rpaths necessarily
+            # come from user variables. The idea of using OrderedSet
```

Comment:
> duplicate rpaths necessarily come from user variables

Why does meson removes stuff explicitly added by the user? I had a similar problem with asan, where meson just messed up the rpath stuff on install, making the installed asan build completely unusable and I had to manually re-add stuff with `install_name_tool`.

Reply:
Because modifying binaries makes for faster building/installing that relinking. On the other hand, in-place editing is more stateful and thus requires trickier handling.

Reply:
I was not questioning why we modify binaries. I was questioning why we remove rpaths that are explicitly added in arguments. IMO meson should only remove rpaths it added itself…

Reply:
This is a long-standing discussion in Meson that I eventually gave up on. I believe we should only remove RPATHs that point to the build directory.

Reply:
I agree with you, everything else is very likely to break things.

Reply:
While an exception to this should be rpaths that are explicitly added only for the build via the respective kwarg.

Reply:
The RPATH deletion is done on install, so there should be no exception to that rule otherwise the resultant binaries are not reproducible.

Reply:
I am not sure what you mean by that, can you clarify?
What I mean is that, if in the meson build file I set a build rpath, I expect that to be removed on install, regardless where it points to:


> `build_rpath` a string to add to target's rpath definition in the build dir, but which will be removed on install

Reply:
>  we should only remove RPATHs that point to the build directory.

This rule wouldn't care what the source of that RPATH is; whether it's `build_rpath:` or `add_project_link_arguments('-Wl,-rpath=' + meson.build_root())`, etc.

We should also remove any paths pointing to the source directory, because that will also result in non-reproducible builds.

Reply:
> This rule wouldn't care what the source of that RPATH is;

That would essentially break the `build_rpath` option, which I think is unacceptable.

Reply:
Just to be clear what I mean: I agree that all rpaths pointing to build or source dir should be removed on install, but it still has to be ensured that all rpaths added with `build_rpath` are removed too, independent of where they are pointing to.

Reply:
Ah, right, I misunderstood what you meant. I agree with you.